pkg/storage/ingest/partition_offset_reader.go

```diff
@@ -0,0 +1,203 @@
+			Help:        "Total number of failed requests to get the last produced offset.",
+			ConstLabels: prometheus.Labels{"partition": strconv.Itoa(int(partitionID))},
+		}),
+		lastProducedOffsetLatency: promauto.With(reg).NewSummary(prometheus.SummaryOpts{
+			Name:        "cortex_ingest_storage_reader_last_produced_offset_request_duration_seconds",
+			Help:        "The duration of requests to fetch the last produced offset of a given partition.",
+			ConstLabels: prometheus.Labels{"partition": strconv.Itoa(int(partitionID))},
+			Objectives:  latencySummaryObjectives,
+			MaxAge:      time.Minute,
+			AgeBuckets:  10,
+		}),
+	}
+
+	p.Service = services.NewTimerService(pollFrequency, nil, p.onPollInterval, p.stopping)
+
+	return p
+}
+
+func (p *partitionOffsetReader) onPollInterval(ctx context.Context) error {
+	// The following call blocks until the last produced offset has been fetched from Kafka. If fetching
+	// the offset takes longer than the poll interval, than we'll poll less frequently than configured.
+	p.getAndNotifyLastProducedOffset(ctx)
+
+	// Never return error, otherwise the service stops.
+	return nil
+}
+
+func (p *partitionOffsetReader) stopping(_ error) error {
+	// Release any waiting goroutine without swapping the result waiter so that if any other goroutine
+	// will watch it after this point it will get immediately notified.
+	p.nextResultWaiterMx.Lock()
+	p.nextResultWaiter.notify(0, errPartitionOffsetReaderStopped)
+	p.nextResultWaiterMx.Unlock()
+
+	return nil
+}
+
+// getAndNotifyLastProducedOffset fetches the last produced offset for a partition and notifies all waiting
+// goroutines (if any).
+func (p *partitionOffsetReader) getAndNotifyLastProducedOffset(ctx context.Context) {
+	// Swap the next waiter with a new one.
+	p.nextResultWaiterMx.Lock()
+	wg := p.nextResultWaiter
+	p.nextResultWaiter = newResultWaiter[int64]()
+	p.nextResultWaiterMx.Unlock()
+
+	// We call getLastProducedOffset() even if there are no goroutines waiting on the result in order to get
+	// a constant load on the Kafka backend. In other words, the load produced on Kafka by this component is
+	// constant, regardless the number of received queries with strong consistency enabled.
+	offset, err := p.getLastProducedOffset(ctx)
+	if err != nil {
+		level.Warn(p.logger).Log("msg", "failed to fetch the last produced offset", "err", err)
+	} else {
+		level.Info(p.logger).Log("msg", "fetched the last produced offset", "offset", offset)
+	}
+
+	// Notify whoever was waiting for it.
+	wg.notify(offset, err)
+}
+
+// getLastProducedOffset fetches and returns the last produced offset for a partition, or -1 if the
+// partition is empty. This function issues a single request, but the Kafka client used under the
+// hood may retry a failed request until the retry timeout is hit.
+func (p *partitionOffsetReader) getLastProducedOffset(ctx context.Context) (_ int64, returnErr error) {
+	startTime := time.Now()
+
+	p.lastProducedOffsetRequestsTotal.Inc()
```

Comment:
Yes, and this is done on purpose to show a change in the metric while the request is running (e.g. think about stuck requests and you see no increase in `requests_total`... it will look weird). We do the exact same in nearly every other place where we track total and failed requests count.

Reply:
ok i suppose consistency is more important here. happy to keep it as-is